modules/azure/compute.go

```diff
@@ -8,97 +8,392 @@ import (
+	return osDiskName
 }
 
-// GetSizeOfVirtualMachineE gets the size type of the given Azure Virtual Machine
-func GetSizeOfVirtualMachineE(t testing.TestingT, vmName string, resGroupName string, subscriptionID string) (compute.VirtualMachineSizeTypes, error) {
-	// Validate resource group name and subscription ID
-	resGroupName, err := getTargetAzureResourceGroupName(resGroupName)
+// GetVirtualMachineOSDiskNameE gets the OS Disk name of the specified Azure Virtual Machine.
+func GetVirtualMachineOSDiskNameE(vmName string, resGroupName string, subscriptionID string) (string, error) {
+	// Get VM Object
+	vm, err := GetVirtualMachineE(vmName, resGroupName, subscriptionID)
 	if err != nil {
 		return "", err
 	}
 
-	// Create a VM client
-	vmClient, err := GetVirtualMachineClientE(subscriptionID)
+	return *vm.StorageProfile.OsDisk.Name, nil
+}
+
+// GetVirtualMachineAvailabilitySetID gets the Availability Set ID of the specified Azure Virtual Machine.
+// This function would fail the test if there is an error.
+func GetVirtualMachineAvailabilitySetID(t testing.TestingT, vmName string, resGroupName string, subscriptionID string) string {
+	avsID, err := GetVirtualMachineAvailabilitySetIDE(vmName, resGroupName, subscriptionID)
+	require.NoError(t, err)
+
+	return avsID
+}
+
+// GetVirtualMachineAvailabilitySetIDE gets the Availability Set ID of the specified Azure Virtual Machine.
+func GetVirtualMachineAvailabilitySetIDE(vmName string, resGroupName string, subscriptionID string) (string, error) {
+	// Get VM Object
+	vm, err := GetVirtualMachineE(vmName, resGroupName, subscriptionID)
 	if err != nil {
 		return "", err
 	}
 
-	// Get the details of the target virtual machine
-	vm, err := vmClient.Get(context.Background(), resGroupName, vmName, compute.InstanceView)
+	// Virtual Machine has no associated Availability Set
+	if vm.AvailabilitySet == nil {
+		return "", nil
+	}
+
+	return GetNameFromResourceID(*vm.AvailabilitySet.ID), nil
+}
+
+// VMImage represents the storage image for the specified Azure Virtual Machine.
+type VMImage struct {
+	Publisher string
+	Offer     string
+	SKU       string
+	Version   string
+}
+
+// GetVirtualMachineImage gets the Image of the specified Azure Virtual Machine.
+// This function would fail the test if there is an error.
+func GetVirtualMachineImage(t testing.TestingT, vmName string, resGroupName string, subscriptionID string) VMImage {
+	vmImage, err := GetVirtualMachineImageE(vmName, resGroupName, subscriptionID)
+	require.NoError(t, err)
+
+	return vmImage
+}
+
+// GetVirtualMachineImageE gets the Image  of the specified Azure Virtual Machine.
+func GetVirtualMachineImageE(vmName string, resGroupName string, subscriptionID string) (VMImage, error) {
+	var vmImage VMImage
+
+	// Get VM Object
+	vm, err := GetVirtualMachineE(vmName, resGroupName, subscriptionID)
+	if err != nil {
+		return vmImage, err
+	}
+
+	if vm.StorageProfile == nil {
+		return vmImage, NewNotFoundError("Image Reference", "Any", vmName)
```

Comment:
Why does a `nil` storage profile mean a `NewNotFoundError`?

Reply:
I checked and this is required as well, removed.